subsys/logging/log_msg.c

```diff
@@ -38,9 +39,25 @@ static void cont_free(struct log_msg_cont *cont)
 
 static void msg_free(struct log_msg *msg)
 {
+	u32_t nargs = msg->hdr.params.std.nargs;
```

Comment:
IMHO you have to scan format string and run this check only for arguments which uses "%s".

Reply:
i disagree. Now, as mechanism protects against double free, the only faulty scenario is following:

```
LOG_INF("%d", flags);
LOG_INF("%s" log_strdup(str));
/* flags == address returned by log_strdup */
```

What will happen here:
First log will find matching address and flag indicating that buffer is allocated. Buffer will be freed.
Second log will not free the buffer (flag is zeroed) or it will free it if log_strdup() was called somewhere else and same buffer was again allocated. In that case %s will print wrong content (newer duplicate). Imo, risk of such situation is very low and outcome of such situation is not critical.
String scanning is not deterministic (variable string length) so im strongly against that.

Reply:
Here is the scenario I am afraid:
```
LOG_INF("0x%08X", flags) // flags = 0x20002008
LOG_INF("%s", log_strdup(str)); // log_strdup(str) = 0x20002000
```
The `log_is_strdup(flags)` will return 1. Then you will corrupt value passed in next log statement using `atomic_dec((atomic_t *)buf) == 1` check, which may succeed freeing the buffer prematurely.

IMHO hidden corruption of data in logger is not acceptable.